
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <title>inmembrane</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta name="description" content="">
    <meta name="author" content="">

    <!-- Le styles -->
    <link href="css/bootstrap.css" rel="stylesheet">
    <style type="text/css">
      body {
        padding-top: 20px;
        padding-bottom: 40px;
      }

      /* Custom container */
      .container-narrow {
        margin: 0 auto;
        max-width: 700px;
      }
      .container-narrow > hr {
        margin: 30px 0;
      }

      /* Main marketing message and sign up button */
      .jumbotron {
        margin: 60px 0;
        text-align: center;
      }
      .jumbotron h1 {
        font-size: 72px;
        line-height: 1;
      }
      .jumbotron .btn {
        font-size: 21px;
        padding: 14px 24px;
      }

      /* Supporting marketing content */
      .marketing {
        margin: 60px 0;
      }
      .marketing p + h4 {
        margin-top: 28px;
      }
    </style>
    <link href="css/bootstrap-responsive.css" rel="stylesheet">

    <!-- HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->

    <!-- Fav and touch icons -->
    <link rel="shortcut icon" href="ico/favicon.ico">
    <link rel="apple-touch-icon-precomposed" sizes="144x144" href="ico/apple-touch-icon-144-precomposed.png">
    <link rel="apple-touch-icon-precomposed" sizes="114x114" href="ico/apple-touch-icon-114-precomposed.png">
    <link rel="apple-touch-icon-precomposed" sizes="72x72" href="ico/apple-touch-icon-72-precomposed.png">
    <link rel="apple-touch-icon-precomposed" href="ico/apple-touch-icon-57-precomposed.png">
  </head>

  <body>

    <div class="container-narrow">

      <div class="masthead">
        <ul class="nav nav-pills pull-right">
          <li><a href="index.html">Home</a></li>
          <li class="active"><a href="#">Customization</a></li>
          <li><a href="api.html">API</a></li>
        </ul>
        <h3 class="muted">inmembrane</h3>
      </div>

      <hr>

      <div class="row-fluid marketing">
        <div class="span12">
         
         
<h1>Customization of <em>inmembrane</em> parameters</h1>
<p>All adjustable parameters of the running of <em>inmembrane</em> have been collected in a parameters file, which is typically stored in the <code>inmembrane.config</code>. On the first use of the program, when no such <code>inmembrane.config</code> is found, a default <code>inmembrane.config</code> will be generated, otherwise existing <code>inmembrane.config</code> will be used.
</p>
<p>The default parameters are as following, they will be explained in sections below:
</p>
<pre><code>{
  'fasta': '',
  'csv': '',
  'out_dir': '',
  'protocol': 'gram_pos', # 'gram_neg'

#### Signal peptide and transmembrane helix prediction
  'signalp4_bin': '',
# 'signalp4_bin': 'signalp',
  'lipop1_bin': '',
# 'lipop1_bin': 'LipoP',
  'tmhmm_bin': '',
#  'tmhmm_bin': 'tmhmm',
  'memsat3_bin': 'runmemsat',
  'helix_programs': ['tmhmm'],
# 'helix_programs': ['tmhmm', 'memsat3'],
  'terminal_exposed_loop_min': 50, # unused in gram_neg protocol
  'internal_exposed_loop_min': 100, # try 30 for gram_neg

#### Sequence similarity and motif prediction
  'hmmsearch3_bin': 'hmmsearch',
  'hmm_profiles_dir': '%(hmm_profiles)s',
  'hmm_evalue_max': 0.1,
  'hmm_score_min': 10,

#### Outer membrane beta-barrel predictors
  'barrel_programs': ['bomp', 'tmbetadisc-rbf'],
# 'barrel_programs': ['bomp', 'tmbetadisc-rbf', 'tmbhunt', 'tmbeta'],
  'bomp_clearly_cutoff': 3, # &gt;= to this, always classify as an OM(barrel)
  'bomp_maybe_cutoff': 1, # must also have a signal peptide to be OM(barrel)
  'tmbhunt_clearly_cutoff': 0.95,
  'tmbhunt_maybe_cutoff': 0.5,
  'tmbetadisc_rbf_method': 'aadp', # aa, dp, aadp or pssm
}
</code></pre><p>Existing <code>inmembrane.config</code> files can be changed. The parameters will be explained below, grouped in terms of related parameters.
</p>

<h3>General Parameters</h3>
<p>These are the general parameters to the program, representing the top-down organization of the work-flow. 
</p>
<p>A fasta program is the input, following one of currently two different protocols - one for Gram+ bacteria, and one for Gram- bacteria. The results are saved to a csv file as an output.
</p>
<ul>
 <li>
     'fasta': the pathname of the fasta file holding the sequences
 </li>

 <li>
     'protocol': switches between different protocols for analysising the proteomes
 </li>

 <li>
     'csv': the name of the output file in CSV (Excel-compatible) format. Defaults to a file using the same filename as 'fasta'
 </li>
</ul>
<p>For potential debugging, it's important that all intermediate files are saved, and they are saved in: 
</p>
<ul>
 <li>
     'out_dir': the directory where all intermediate output files are located
 </li>
</ul>

<h3>Optional - Binaries or Web-based API's</h3>
<p>One of the problems that <em>inmembrane</em> tackles is the complexity of using and installing multiple binaries. In older programs, the only option was to install local binaires of the external programs. This raises problems involved with getting the binary for the right platform, compilation with the right tools, and installation in the correct file system. In contrast, many of these tools now provide a web-interface. In <em>inmembrane</em>, we have added facilities to use the web service, whether through a RESTful interface, or through a SOAP interface, when provided.
</p>
<p>Consequently, the parameters for the binaries are needed, where <em>inmembrane</em> will attempt to run the program in the default directory, or if a full-path name is given, <em>inmembrane</em> can run the binary in the correct location. 
   In order to turn off the local binary option, if in a *_bin parameter is given the empty string, the local binary step will be skipped and the web version will be used instead.
</p>
<ul>
 <li>
     'signalp4_bin', 'lipop1_bin', 'tmhmm_bin', 'memsat3_bin': the full pathname of the binaries used for the analysis. If empty, it is not run, and the web version is used instead
 </li>
</ul>

<h3>The Surface-Exposure Algorithm</h3>
<p>The key part of the surface-exposure prediction evaluates the length of loops in the extracellular part of membrane proteins. This evaluation requires the key parameter of loop cutoff that determines the minimum length of a loop that is required to potentially pass through the glycan layer. 
</p>
<p>In the original <em>SurfG+</em> paper, for Gram+ bacteria, it was found that a loop lenght of ~50 amino acids is required. Therefore for internal loops, this gives a length ~100 amino acids, which allows the loop to pass through the layer and back down again.
</p>
<ul>
 <li>
     'terminal_exposed_loop_min': this is the length in terms of amino acids given to define an external loop of a protein to be long enough to stick out of the bacterial cell-wall to be considered surface-epxosed
 </li>

 <li>
     'internal_exposed_loop_min': similarly to above, except is double the length as internal loops need to go up, and down again, for continuity with the rest of the protein.
 </li>
</ul>
<p>There are several different way to identify proteins that have loops that potentially stick out of the glycan cell-wall of the bacteria. These will be discussed in the following sections.
</p>

<h3>Proteins attached to the Glycan Cell-wall and the Lipid bilayer</h3>
<p>The clearest example of proteins that may stick of the glycan layer are proteins that possess known cell-wall attachment motifs. 
</p>
<p>Motifs for motifs that bind to the glycan cell-wall are taken from PFAM as HMMER profiles, and stored into a directory included in the <em>inmembrane</em> package. In the default <code>inmembrane.config</code> file, the location of this directory will be put in <code>hmm_profiles_dir</code>. 
</p>
<p>Subsequently, <code>HMMER</code> will be used to search the profiles against the input query sequences. The cutoffs for the identification of a motif is given by <code>hmm_evalue_max</code> and <code>hmm_score_min</code>.
</p>
<ul>
 <li>
     'hmmsearch3_bin': the binary for HMMER
 </li>

 <li>
     'hmm_profiles_dir': the directory that holds the HMMER profiles for cell-wall signals
 </li>

 <li>
     'hmm_evalue_max': the maximum expectation value cutoff for acceptable prediction in HMMER
 </li>

 <li>
     'hmm_score_min': the minimum score cutoff for acceptatble prediction in HMMER
 </li>
</ul>
<p>Another source of proteins attached to the membrane are lipo-proteins. LipoP is used to identify motifs for the cell to attach a protein to the lipid layer of the protein. 
</p>
<ul>
 <li>
     'lipop_bin': the location of the binary of LipoP, or if this is given as '', then it uses the web-version at http://http://www.cbs.dtu.dk/services/LipoP/.
 </li>
</ul>

<h3>Transmembrane-&alpha;-helix Predictors</h3>
<p>The other main source of proteins in Gram+ proteins with external loops are the identification of proteins with transmembrane &amp;alpha-helices. The internal loops of proteins with multiple transmembrane helices, as well as the terminal loops are all candidates to be exposed out of the glycan layer.
</p>
<p>There is no dominant transmembrane helix predictor, and thus <em>inmembrane</em> has been written to allow pluggable options for transmembrane helix prediction. To allow for this, <em>inmembrane</em> allows the results of more than one helix predictor, and the one that give the longest loops that extends out of the extracellular region will be used. As such the parameter 'helix_programs' lists all the transmembrane helix predictors to be used. Currently, 
</p>
<ul>
 <li>
     'helix_programs': a list of transmembrane-helix predictors to be used. Currently, <em>inmembrane</em> can handle <code>tmhmm</code> and <code>memsat</code>.
 </li>
</ul>

<h3>&beta;-barrel predictors</h3>
<ul>
 <li>
     'barrel_programs': list of programs to do &amp;beta-barel ['bomp', 'tmbetadisc-rbf'],
 </li>

 <li>
     'bomp_clearly_cutoff': 3, # &gt;= to this, always classify as an OM(barrel)
 </li>

 <li>
     'bomp_maybe_cutoff': 1, # must also have a signal peptide to be OM(barrel)
 </li>

 <li>
     'tmbhunt_clearly_cutoff': 0.95,
 </li>

 <li>
     'tmbhunt_maybe_cutoff': 0.5,
 </li>

 <li>
     'tmbetadisc_rbf_method': 'aadp', # aa, dp, aadp or pssm
 </li>
</ul>


        </div>
      </div>

      <hr>

      <div class="footer">
        <p>&copy; 2012, Andrew Perry and Bosco Ho</p>
      </div>

    </div> <!-- /container -->

    <!-- Le javascript
    ================================================== -->
    <!-- Placed at the end of the document so the pages load faster -->
    <script src="js/jquery.js"></script>
    <script src="js/bootstrap-transition.js"></script>
    <script src="js/bootstrap-alert.js"></script>
    <script src="js/bootstrap-modal.js"></script>
    <script src="js/bootstrap-dropdown.js"></script>
    <script src="js/bootstrap-scrollspy.js"></script>
    <script src="js/bootstrap-tab.js"></script>
    <script src="js/bootstrap-tooltip.js"></script>
    <script src="js/bootstrap-popover.js"></script>
    <script src="js/bootstrap-button.js"></script>
    <script src="js/bootstrap-collapse.js"></script>
    <script src="js/bootstrap-carousel.js"></script>
    <script src="js/bootstrap-typeahead.js"></script>

  </body>
</html>
