
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <title>inmembrane</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta name="description" content="">
    <meta name="author" content="">

    <!-- Le styles -->
    <link href="css/bootstrap.css" rel="stylesheet">
    <style type="text/css">
      body {
        padding-top: 20px;
        padding-bottom: 40px;
      }

      /* Custom container */
      .container-narrow {
        margin: 0 auto;
        max-width: 700px;
      }
      .container-narrow > hr {
        margin: 30px 0;
      }

      /* Main marketing message and sign up button */
      .jumbotron {
        margin: 60px 0;
        text-align: center;
      }
      .jumbotron h1 {
        font-size: 72px;
        line-height: 1;
      }
      .jumbotron .btn {
        font-size: 21px;
        padding: 14px 24px;
      }

      /* Supporting marketing content */
      .marketing {
        margin: 60px 0;
      }
      .marketing p + h4 {
        margin-top: 28px;
      }
    </style>
    <link href="css/bootstrap-responsive.css" rel="stylesheet">

    <!-- HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->

    <!-- Fav and touch icons -->
    <link rel="shortcut icon" href="ico/favicon.ico">
    <link rel="apple-touch-icon-precomposed" sizes="144x144" href="ico/apple-touch-icon-144-precomposed.png">
    <link rel="apple-touch-icon-precomposed" sizes="114x114" href="ico/apple-touch-icon-114-precomposed.png">
    <link rel="apple-touch-icon-precomposed" sizes="72x72" href="ico/apple-touch-icon-72-precomposed.png">
    <link rel="apple-touch-icon-precomposed" href="ico/apple-touch-icon-57-precomposed.png">
  </head>

  <body>

    <div class="container-narrow">

      <div class="masthead">
        <ul class="nav nav-pills pull-right">
          <li class="active"><a href="#">Home</a></li>
          <li><a href="#">About</a></li>
        </ul>
        <h3 class="muted">inmembrane</h3>
      </div>

      <hr>

      <div class="jumbotron">
        <h1>inmembrane</h1>
        <p class="lead">Prediction of surface-exposed proteins in bacterial cells</p>
        <a class="btn btn-large btn-success" href="#">Download</a>
      </div>

      <hr>

      <div class="row-fluid marketing">
        <div class="span12">
<h2>Getting it to Work</h2>

<h4>What is <em>inmembrane</em>?</h4>
<p><em>inmembrane</em> is a program to predict if a protein is likely to be surface-exposed in a bacterial cell. It is designed to be relatively easy to install, with several different options to help you run all external binaries on your system.
</p>

<h4>How can we help you?</h4>
<p>We envisage <em>inmembrane</em> being of use in three ways:
</p>
<ol>
 <li><p>You want to do a quirk and dirty annotation of your proteome for cell-surface exposure in order to cross-check your experiment. Keep reading on this page.
</p>

 </li>

 <li><p>You want to play around with different ways of cell localization assignment. Check out our parameterization guide.
</p>

 </li>

 <li><p>You want to make your own bioinformatic work-flow, and would like to use <em>inmembrane</em> as a useful Python starting point. Have a look at our programming model, paper, and API guide.
</p>

 </li>
</ol>

<h4>Combining analyses into a workflow</h4>
<p>The cell-surface exposure algorithm is taken from the paper (Barinove et al. 2009. Proteomics 9:61-73). It is a straightforward case of a simple work-flow defined over a number of external binaries that do:
</p>
<ul>
 <li>
     transmembrane helix prediction,
 </li>

 <li>
     lipoprotein prediction,
 </li>

 <li>
     secretion signal prediction
 </li>

 <li>
     cell-wall binding motif prediction
 </li>
</ul>
<p>Although there exists several generic bioinformatic workflow packages, the deduction of cell-surface exposure from the results of these programs is a little more involved. Here, we have written this in a Python package, and in a manner that is extensible, and potentially repackageable for other bioinformatic analysis. 
</p>

<h4>Ease of use</h4>
<p>We have specifically chosen Python for the overall ease of use, and readability. Read our paper for more details. One of the most difficult aspects of installing a workflow like this is the installation of multiple external binaries. We have made a large effort to incorporate web-based modules, which in most cases, should significantly simplify the process of installation.
</p>

<h4>How to install?</h4>
<p>Well, to install the program, you'll have to have install
</p>
<ul>
 <li>
     Python: it should be on most systems, but if not, go to <a href="http://www.python.org">http://www.python.org</a>
 </li>

 <li>
     pip: <a href="http://pypi.python.org/pypi/pip">http://pypi.python.org/pypi/pip</a>
 </li>

 <li>
     HMMER: this is the only bioinformatic software that definitely needs to be installed <a href="http://hmmer.janelia.org/software">http://hmmer.janelia.org/software</a>
 </li>
</ul>
<p>Once these are installed on your system, you should be able to do this to install <em>inmembrane</em>:
</p>
<pre><code>&gt;&gt;&gt; pip install inmembrane
</code></pre>
<h4>Running <em>inmembrane</em></h4>
<p>Let's say you have all the protein sequences you want to analyze in a FASTA text file called <code>proteome.fasta</code>.
</p>
<p>Inmembrane can be run from the command line, assuming you have an internet connection, and all the web-sites <em>inmembrane</em> relies on are up:
</p>
<pre><code>&gt;&gt;&gt; inmembrane_scan proteome.fasta
</code></pre><p>The results will be written to proteome.csv, which can be opened by Microsoft Excel. 
</p>

<h4>Interpreting output</h4>
<p>It should look something like this: 
</p>
<pre><code>SPy_0008  CYTOPLASM(non-PSE)  .                         SPy_0008 from AE004092
SPy_0010  PSE-Membrane        tmhmm(1)                  SPy_0010 from AE004092
SPy_0012  PSE-Cellwall        hmm(GW2|GW3|GW1);signalp  SPy_0012 from AE004092
SPy_0016  MEMBRANE(non-PSE)   tmhmm(12)                 SPy_0016 from AE004092
SPy_0019  SECRETED            signalp                   SPy_0019 from AE004092
</code></pre><p>There are 4 columns to the result: 
</p>
<ol>
 <li>
     The sequence identifier (seqid) is the first word in the id line of the FASTA file
 </li>

 <li>
     The cell-localization category:<ul>
 <li>
     PSE: means Potentially Surface Exposed, this can arise from<ul>
 <li>
     transmembrane helical protein with long longs
 </li>

 <li>
     lipoprotein attached to the lipid layer
 </li>

 <li>
     protein with motif that attaches to the glycan cell-wall
 </li>
</ul>

 </li>

 <li>
     SECRETED: contains a secretion signal
 </li>

 <li>
     CYTOPLASM: basically none of the above
 </li>
</ul>

 </li>

 <li>
     Summary of the results from the external bioinformatic programs. In this case, we have hits from <code>tmhmm</code>, <code>hmm</code>, and <code>signalp</code>
 </li>

 <li>
     The description of the protein taken from the FASTA file.
 </li>
</ol>
<hr />


<h2>Messing around with the parameters</h2>

<h4>The parameters in the program</h4>
<p>All adjustable parameters of the running of <em>inmembrane</em> have been collected in a parameters file, which is typically stored in the <code>inmembrane.config</code>. 
</p>
<p>On the first use of the program, when no such <code>inmembrane.config</code> is found, a default <code>inmembrane.config</code> will be generated. Existing <code>inmembrane.config</code> will be read and interpreted.
</p>

<h3>General parameters</h3>
<p>These are the general parameters to the program, representing the top-down organization of the work-flow. 
</p>
<p>A fasta program is the input, following one of currently two different protocols - one for Gram+ bacteria, and one for Gram- bacteria. The results are saved to a csv file as an output.
</p>
<ul>
 <li>
     'fasta': the pathname of the fasta file holding the sequences
 </li>

 <li>
     'protocol': switches between different protocols for analysising the proteomes
 </li>

 <li>
     'csv': the name of the output file in CSV (Excel-compatible) format. Defaults to a file using the same filename as 'fasta'
 </li>
</ul>
<p>For potential debugging, it's important that all intermediate files are saved, and they are saved in: 
</p>
<ul>
 <li>
     'out_dir': the directory where all intermediate output files are located
 </li>
</ul>

<h3>Optional binaries</h3>
<ul>
 <li>
     'signalp4_bin', 'lipop1_bin', 'tmhmm_bin', 'memsat3_bin': the full pathname of the binaries used for the analysis. If empty, it is not run, and the web version is used instead
 </li>
</ul>

<h4>transmembrane helix predictors</h4>
<ul>
 <li>
     'helix_programs': a choice of which transmembrane-helix predictors are used. Currently, the program loops over all designated helix predictors and chooses the longest predicted loops.
 </li>
</ul>

<h4>key parameters to determine surface-exposure</h4>
<ul>
 <li>
     'terminal_exposed_loop_min': this is the length in terms of amino acids given to define an external loop of a protein to be long enough to stick out of the bacterial cell-wall to be considered surface-epxosed
 </li>

 <li>
     'internal_exposed_loop_min': similarly to above, except is double the length as internal loops need to go up, and down again, for continuity with the rest of the protein.
 </li>
</ul>

<h4>HMMER parameters for signal proteins for cell-wall attachment</h4>
<ul>
 <li>
     'hmmsearch3_bin': the binary for HMMER
 </li>

 <li>
     'hmm_profiles_dir': the directory that holds the HMMER profiles for cell-wall signals
 </li>

 <li>
     'hmm_evalue_max': the maximum expectation value cutoff for acceptable prediction in HMMER
 </li>

 <li>
     'hmm_score_min': the minimum score cutoff for acceptatble prediction in HMMER
 </li>
</ul>

<h4>&beta;-barrel predictors</h4>
<ul>
 <li>
     'barrel_programs': list of programs to do &amp;beta-barel ['bomp', 'tmbetadisc-rbf'],
 </li>

 <li>
     'bomp_clearly_cutoff': 3, # &gt;= to this, always classify as an OM(barrel)
 </li>

 <li>
     'bomp_maybe_cutoff': 1, # must also have a signal peptide to be OM(barrel)
 </li>

 <li>
     'tmbhunt_clearly_cutoff': 0.95,
 </li>

 <li>
     'tmbhunt_maybe_cutoff': 0.5,
 </li>

 <li>
     'tmbetadisc_rbf_method': 'aadp', # aa, dp, aadp or pssm
 </li>
</ul>
<hr />


<h2>Programming philosophy</h2>
        </div>
      </div>

      <hr>

      <div class="footer">
        <p>&copy; 2012, Andrew Perry and Bosco Ho</p>
      </div>

    </div> <!-- /container -->

    <!-- Le javascript
    ================================================== -->
    <!-- Placed at the end of the document so the pages load faster -->
    <script src="js/jquery.js"></script>
    <script src="js/bootstrap-transition.js"></script>
    <script src="js/bootstrap-alert.js"></script>
    <script src="js/bootstrap-modal.js"></script>
    <script src="js/bootstrap-dropdown.js"></script>
    <script src="js/bootstrap-scrollspy.js"></script>
    <script src="js/bootstrap-tab.js"></script>
    <script src="js/bootstrap-tooltip.js"></script>
    <script src="js/bootstrap-popover.js"></script>
    <script src="js/bootstrap-button.js"></script>
    <script src="js/bootstrap-collapse.js"></script>
    <script src="js/bootstrap-carousel.js"></script>
    <script src="js/bootstrap-typeahead.js"></script>

  </body>
</html>
